### PR TITLE
Add ru.proninyaroslav.libretrack

### PR DESCRIPTION
**Pull request by proninyaroslav**
### Please confirm your submission meets all the criteria

- [x] I have read the [App Requirements][reqs] and [App Maintenance][maint] pages.
- [x] My pull request follows the instructions at [App Submission][submission].
- [ ] I am using only the minimal set of permissions. *(If not, please explain each non-standard permission.)*
 - `--persist=.dart_tool`: required for storing Floor database https://pub.dev/packages/floor. Proposing a change on the XDG path https://github.com/vitusortner/floor/issues/593
- [x] All assets referenced in the manifest are redistributable by any party.  If not, the unredistributable parts are using an extra-data source type.
- [x] I am an upstream contributor to the project. If not, I contacted upstream developers about submitting their software to Flathub. **Link:**
- [x] I own the domain used in the application ID or the domain has a policy for delegating subdomains (e.g. GitHub, SourceForge).
- [x] Any additional patches or files have been submitted to the upstream projects concerned. *(If not, explain why.)*

[reqs]: https://github.com/flathub/flathub/wiki/App-Requirements
[maint]: https://github.com/flathub/flathub/wiki/App-Maintenance
[submission]: https://github.com/flathub/flathub/wiki/App-Submission

### CODE REVIEW

File: flathub.json [hunk 1]
```diff
@@ -0,0 +1,6 @@
+{
+  "only-arches": [
```
Comment (hfiguiere) [inline flathub.json:2]: you don't need this these are the only arch supported today.

File: ru.proninyaroslav.libretrack.metainfo.xml [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+        <screenshot>
+            <image type="source">https://github.com/proninyaroslav/libretrack/blob/master/art/linux_screenshots/s7.png</image>
+        </screenshot>
+    </screenshots>
+    <launchable type="desktop-id">ru.proninyaroslav.libretrack.desktop</launchable>
+    <releases>
+        <release version="1.2.0" date="2021-08-12"/>
+    </releases>
+    <update_contact>proninyaroslav@gmail.com</update_contact>
+</component>
```
Comment (hfiguiere) [inline ru.proninyaroslav.libretrack.metainfo.xml:65]: it's missing the oars rating. (required)

File: ru.proninyaroslav.libretrack.yaml [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+finish-args:
+  # Required due to being a GUI application
+  - --socket=x11
+  - --socket=fallback-x11
+  # Required to make sure x11 performance is achived on all platforms
+  # At least that's what the legends tell. it might be worth experimenting
+  # with dropping this permission.
+  - --share=ipc
+  # Required for experimental wayland support
+  - --socket=wayland
```
Comment (hfiguiere) [inline ru.proninyaroslav.libretrack.yaml:16]: it's either `wayland` + `fallback-x11` or just `x11`.

File: ru.proninyaroslav.libretrack.yaml [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+  # Required for notifications in various desktop environments
+  - --talk-name=org.freedesktop.Notifications
+  - --talk-name=org.kde.StatusNotifierWatcher
+  # Required due to KDE's special way of handling notifications
+  - --own-name=org.kde.StatusNotifierItem-2-1
+  # Required for advanced input methods e.g. writing CJK languages
+  - --talk-name=org.freedesktop.portal.Fcitx
+  # Required for correct work of Flutter packages, which use system DBus
+  # for example https://pub.dev/packages/connectivity_plus
+  - --socket=system-bus
```
Comment (hfiguiere) [inline ru.proninyaroslav.libretrack.yaml:31]: can't you just allow the proper dbus names?

File: ru.proninyaroslav.libretrack.yaml [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+  # TODO: Temporary fix https://github.com/canonical/dbus.dart/issues/257
+  - --socket=session-bus
```
Comment (hfiguiere) [inline ru.proninyaroslav.libretrack.yaml:33]: and here too?

Wildcard are allowed.

`--log-session-bus` and `--log-system-bus` are usefull to figure it out.

File: ru.proninyaroslav.libretrack.yaml [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+      - -DBUILD_SHARED_LIBS=ON
+    sources:
+      - type: archive
+        url: https://github.com/open-source-parsers/jsoncpp/archive/refs/tags/1.7.5.tar.gz
+        sha256: 4338c6cab8af8dee6cdfd54e6218bd0533785f552c6162bb083f8dd28bf8fbbe
+  - name: libretrack
+    buildsystem: simple
+    only-arches:
+      - x86_64
+      - aarch64
```
Comment (hfiguiere) [inline ru.proninyaroslav.libretrack.yaml:57]:
```suggestion
```

File: ru.proninyaroslav.libretrack.yaml [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+    build-commands:
+      - mkdir -p /app
+      - cp -r libretrack /app/libretrack
+      - chmod +x /app/libretrack/libretrack
+      - install -Dm644 /app/libretrack/data/icons/app-icon.svg /app/share/icons/hicolor/scalable/apps/ru.proninyaroslav.libretrack.svg
+      - mkdir -p /app/bin /app/lib
+      - ln -s /usr/lib/x86_64-linux-gnu/libsqlite3.so.0 /app/lib/libsqlite3.so
```
Comment (hfiguiere) [inline ru.proninyaroslav.libretrack.yaml:64]: this will fail on aarch64.
Also why do you need it?

File: ru.proninyaroslav.libretrack.yaml [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+      - ln -s /app/libretrack/libretrack /app/bin/libretrack
+      - install -Dm644 ru.proninyaroslav.libretrack.desktop /app/share/applications/ru.proninyaroslav.libretrack.desktop
+      - install -Dm644 ru.proninyaroslav.libretrack.metainfo.xml /app/share/appdata/ru.proninyaroslav.libretrack.metainfo.xml
+      - install libretrack.sh /app/bin/libretrack
+    sources:
+      - type: archive
+        only-arches:
+          - x86_64
+        url: https://github.com/proninyaroslav/libretrack/releases/download/1.2.0/libretrack-1.2.0-linux-x86_64.tar.gz
```
Comment (hfiguiere) [inline ru.proninyaroslav.libretrack.yaml:73]: can't you build from source?

**Review by hfiguiere — CHANGES_REQUESTED** [review]

File: flathub.json (hunk 1 above)
Comment (proninyaroslav) [reply flathub.json:2]: In other words, arm64 is not supported?

File: ru.proninyaroslav.libretrack.yaml (hunk 4 above)
Comment (proninyaroslav) [reply ru.proninyaroslav.libretrack.yaml:31]: No, because `dbus` package tries to open the system socket file physically and fails if it's not available.

File: ru.proninyaroslav.libretrack.yaml (hunk 5 above)
Comment (proninyaroslav) [reply ru.proninyaroslav.libretrack.yaml:33]: I have given the reason why I can't use `talk-name` at the moment https://github.com/canonical/dbus.dart/issues/257

File: ru.proninyaroslav.libretrack.yaml (hunk 7 above)
Comment (proninyaroslav) [reply ru.proninyaroslav.libretrack.yaml:64]: `libsqlite3` is not automatically linked to the `/lib` dir. At runtime, I will get an exception because the app can't find the library. I will add a check.

File: ru.proninyaroslav.libretrack.yaml (hunk 8 above)
Comment (proninyaroslav) [reply ru.proninyaroslav.libretrack.yaml:73]: Build requires internet connection as I get Dart packages from pub.dev as well as some Linux-specific dependencies for Flutter. In addition, I need to download Flutter itself. As far as I know, Flatpak build system does not allow internet connection during project build, only to get dependencies.

File: ru.proninyaroslav.libretrack.yaml (hunk 4 above)
Comment (proninyaroslav) [reply ru.proninyaroslav.libretrack.yaml:30]: https://github.com/canonical/dbus.dart/blob/main/lib/src/dbus_client.dart#L213

Comment (proninyaroslav) [reply ru.proninyaroslav.libretrack.yaml:30]: I'm getting this error:

`flutter: SocketException: Connection failed (OS Error: No such file or directory, errno = 2), address = /run/dbus/system_bus_socket, port = 0`

Comment (proninyaroslav) [reply ru.proninyaroslav.libretrack.yaml:30]: First of all, they try to get `DBUS_SYSTEM_BUS_ADDRESS` env variable, and if it's not set then `unix:path=/run/dbus/system_bus_socket` is used.

File: flathub.json (hunk 1 above)
Comment (hfiguiere) [reply flathub.json]: only `aarch64` and `x86_64` so its not needed to say they are the only one.

(`arm64` is `aarch64` in case it wasn't clear)

File: ru.proninyaroslav.libretrack.yaml (hunk 8 above)
Comment (hfiguiere) [reply ru.proninyaroslav.libretrack.yaml:70]: fair enough. :-(

**Comment by hfiguiere** [comment]
bot, build ru.proninyaroslav.libretrack

File: ru.proninyaroslav.libretrack.yaml [hunk 9]
```diff
@@ -0,0 +1,96 @@
+app-id: ru.proninyaroslav.libretrack
+runtime: org.gnome.Platform
```
Comment (hfiguiere) [inline ru.proninyaroslav.libretrack.yaml:2]: btw do you really need the GNOME SDK/Platform?

**Comment by proninyaroslav** [comment]
@hfiguiere 
Hmm, what could be wrong with the screenshots? As I understand it, the problem is in the presence of the alpha channel?

**Comment by proninyaroslav** [comment]
@hfiguiere 
I removed the alpha channel but it didn't help. Are there any size or aspect ratio restrictions? I haven't found any documentation on this matter.

**Comment by hfiguiere** [comment]
```
$wget https://github.com/proninyaroslav/libretrack/blob/master/art/linux_screenshots/s1.png
[...]
$ file s1.png 
s1.png: HTML document, UTF-8 Unicode text, with very long lines
```

**Comment by hfiguiere** [comment]
tl;dr this doesn't return the PNG.

**Comment by proninyaroslav** [comment]
Yes, my mistake, I forgot to specify `raw`. Thank you.

**Comment by proninyaroslav** [comment]
Fixed.

**Comment by proninyaroslav** [comment]
bot, build ru.proninyaroslav.libretrack

File: ru.proninyaroslav.libretrack.yaml (hunk 8 above)
Comment (proninyaroslav) [reply ru.proninyaroslav.libretrack.yaml:70]: I have automated the build, so it shouldn't cause any problems in the future https://github.com/proninyaroslav/libretrack/blob/master/.github/workflows/flutter_release_build.yml

**Comment by proninyaroslav** [comment]
@hfiguiere 
I updated the tarballs and I removed the `session-bus` option which was used as a temporary solution.

**Review by hfiguiere — CHANGES_REQUESTED** [review]

File: ru.proninyaroslav.libretrack.yaml (hunk 4 above)
Comment (proninyaroslav) [reply ru.proninyaroslav.libretrack.yaml:30]: I opened an issue on this problem, but it will be fixed in the future https://github.com/canonical/dbus.dart/issues/260

**Comment by nedrichards** [comment]
bot, build ru.proninyaroslav.libretrack

**Comment by nedrichards** [comment]
I see there's a new release of the dbus package that fixes the issue you reported. Will that allow you to remove the broad permission? That's the only thing standing in the way of merging for me.

**Comment by proninyaroslav** [comment]
@nedrichards 
Yes, it allowed to remove `session-bus`. But `system-bus` still remains https://github.com/canonical/dbus.dart/issues/260. Maybe there is no solution to this problem, I don't know for sure.

**Comment by barthalion** [comment]
I posted a comment there.

**Comment by barthalion** [comment]
@proninyaroslav You can track all DBus activity with `flatpak run --log-system-bus`. Blanket access to the system bus is a no-go without a valid reason, and apparently you are well aware of what may be required to make it work.

**Comment by proninyaroslav** [comment]
@barthalion 
I understand. As I said earlier, I'm using the `connectivity_plus` package to check the Internet connection https://github.com/proninyaroslav/libretrack/blob/1.2.0/lib/core/work_manager/fallback_work_manager.dart#L267
This package, like several others, uses the system bus https://github.com/fluttercommunity/plus_plugins/blob/package_info_plus_linux-v1.0.3/packages/connectivity_plus/connectivity_plus_linux/lib/src/network_manager.dart#L20
Could this be a valid reason?

**Comment by barthalion** [comment]
You can specify which system D-Bus names should be allowed with `--system-talk-name=` (it can be passed multiple times). That should make it work.

**Comment by proninyaroslav** [comment]
@barthalion 
Yes, I tried this, but it still threw an exception. Now I checked it again and it seems `dbus` has fixed it. I'll commit it.

**Comment by proninyaroslav** [comment]
bot, build ru.proninyaroslav.libretrack

**Comment by proninyaroslav** [comment]
@barthalion  @nedrichards
Currently `system-bus` has been removed and only `--persist=.dart_tool` remains from non-standard permissions. Maybe there is another way of access, but this is the only one that I found in the documentation.
